File: MatchModule_New/AI/States/ChaceState.cs
using System;
using Games.NB.Match.AI.Actions;
using Games.NB.Match.Base.Attributes;
using Games.NB.Match.Base.Interface;

namespace Games.NB.Match.AI.States
{
    /// <summary>
    /// 跑动状态
    /// </summary>
    [Singleton]
    public class ChaceState : State
    {
        /// <summary>
        /// Represents the instance of the <see cref="ChaceState"/> class.
        /// </summary>
        public static ChaceState Instance
        {
            get { return _state; }
        }

        /// <summary>
        /// Outputs the ChaceState's name.
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return "ChaceState";
        }

        /// <summary>
        /// Initializes the new instance of the <see cref="ChaceState"/> class.
        /// </summary>
        public override void Initialize()
        {
            this.StateChain.Add(this);
            this.StateChain.Add(IdleState.Instance);
            this.StateChain.Add(ActionState.Instance);

            this.StateCondition.Add(ActionState.Instance, ValidateChaceToAction);
            this.StateCondition.Add(this, ValidateChaceToChace);
            this.StateCondition.Add(IdleState.Instance, ValidateChaceToIdle);
        }

        /// <summary>
        /// Chace State action.
        /// </summary>
        /// <param name="player"><see cref="IPlayer"/></param>
        public override void Action(IPlayer player)
        {
            ChaceAction.Instance.Action(player);
        }

        /// <summary>
        /// Decides the nex <see cref="IState"/>
        /// </summary>
        /// <param name="player">Represents the current <see cref="IPlayer"/>.</param>
        /// <param name="preview">Represents the preview <see cref="IState"/></param>
        /// <returns></returns>
        public override IState QuickDecide(IPlayer player, IState preview)
        {
            if (player.Status.NeedRedecide)
            {
                if (player.Status.IsAttackSide == false) // 防守方
                {
                    if (player.Input.AsPosition == Base.Enum.Position.Goalkeeper)
                    {
                        return OffBallState.Instance;
                    }

                    if (player.Status.DefenceStatus.DefenceTarget == null) // 没有防守对象
                    {
                        return PositionalState.Instance;
                    }
                }
                else // 进攻方
                {
                    if (player.Status.Hasball == false)
                    {
                        var ballHandler=player.Match.Status.BallHandler;
                        if (player.Input.AsPosition == Base.Enum.Position.Goalkeeper
                            && null != ballHandler
                            && ballHandler.Status.State is Shoot.RebelShootState)
                            return OffBallState.Instance;
                        return PositionalState.Instance;
                    }
                }
                return ActionState.Instance;
            }
            if (player.Status.DestinationDistanceZero)
                return IdleState.Instance;
            var playerPos = player.Status.Current;
            if (!player.Status.ActiveRegion.IsCoordinateInRegion(playerPos))
                return ChaceState.Instance;
            var ballPos = player.Match.Football.Current;
            if (Math.Abs(playerPos.X - ballPos.X) <= 40 || Math.Abs(playerPos.Y - ballPos.Y) <= 30)
                return ChaceState.Instance;
            return WalkState.Instance;
        }

        #region encapsulation

        private static readonly ChaceState _state = new ChaceState();

        private ChaceState()
        {
            this.Stopable = true;
        }

        private static bool ValidateChaceToAction(IPlayer player, IState preview)
        {
            return player.Status.NeedRedecide;
        }

        private static bool ValidateChaceToChace(IPlayer player, IState preview)
        {
            if (player.Status.NeedRedecide)
            {
                return false;
            }

            if (player.Status.DestinationDistance == 0)
            {
                return false;
            }

            return true;
        }

        private static bool ValidateChaceToIdle(IPlayer player, IState preview)
        {
            if (player.Status.NeedRedecide)
            {
                return false;
            }

            if (player.Status.DestinationDistance > 0)
            {
                return false;
            }

            return true;
        }

        #endregion
    }
}

File: SVN球场AI/HB_SNS/AI/States/ChaceState.cpp
/********************************************************************************
 * 文件名：ChaceState.cpp
 * 创建人：宋斌
 * 创建时间：2011-6-10
 * 版本：1.0
 * 本文件版本号：1.0
 * 最后更新：
 * 功能说明：
 * 历史修改记录：
 * <author>  <time>           <version >   <desc>
 *********************************************************************************/
#include "ChaceState.h"
#include "IdleState.h"
#include "ActionState.h"
#include "OffBallState.h"
#include "PositionalState.h"

#include "../Actions/ChaceAction.h"

ChaceState::ChaceState()
{
    m_Stopable = true;
}

ChaceState::~ChaceState()
{

}

ChaceState* ChaceState::Instance()
{
    static ChaceState instance;

    return &instance;
}

string ChaceState::ToString()
{
    return "ChaceState";
}

void ChaceState::Initialize()
{
    m_StateChain.push_back(this);
    m_StateChain.push_back(IdleState::Instance());
    m_StateChain.push_back(ActionState::Instance());

    m_StateCondition[ActionState::Instance()]       = ValidateChaceToAction;
    m_StateCondition[this]                          = ValidateChaceToChace;
    m_StateCondition[IdleState::Instance()]         = ValidateChaceToIdle;
}

void ChaceState::Action(IPlayer* player)
{
    ChaceAction::Instance()->Action(player);
}

IState* ChaceState::QuickDecide(IPlayer* player, IState* preview)
{
    if (player->Status()->NeedRedecide())
    {
        if (player->Status()->IsAttackSide() == false) // 防守方
        {
            if (player->BuildinAttribute()->GetPosition() == Position_Goalkeeper)
            {
                return OffBallState::Instance();
            }

            if (player->Status()->GetDefenceStatus()->DefenceTarget() == NULL) // 没有防守对象
            {
                return PositionalState::Instance();
            }
        }
        else // 进攻方
        {
            if (player->Status()->Hasball() == false)
            {
                return PositionalState::Instance();
            }
        }

        return ActionState::Instance();
    }
    else
    {
        if (MATH::FloatEqual(player->Status()->DestinationDistance() ,0))
        {
            return IdleState::Instance();
        }
        else
        {
            return ChaceState::Instance();
        }
    }
}

bool ChaceState::ValidateChaceToAction(IPlayer* player, IState* preview)
{
    return player->Status()->NeedRedecide();
}

bool ChaceState::ValidateChaceToChace(IPlayer* player, IState* preview)
{
    if (player->Status()->NeedRedecide())
    {
        return false;
    }

    if (MATH::FloatEqual(player->Status()->DestinationDistance() , 0.0))
    {
        return false;
    }

    return true;
}

bool ChaceState::ValidateChaceToIdle(IPlayer* player, IState* preview)
{
    if (player->Status()->NeedRedecide())
    {
        return false;
    }

    if (player->Status()->DestinationDistance() > 0)
    {
        return false;
    }

    return true;
}

bool ChaceState::IsInState(IState* st)
{
    if (typeid(*this) == typeid(*st)) 
    {
        return true;
    }

    return false;
}